feat(lambda): adding mcp server prompts support to MCPLambdaHandler

Dan-Arnin opened the pull request:
## Summary

This PR adds **prompt registration and rendering functionality** to the `MCPLambdaHandler`.

Previously, MCP servers running on AWS Lambda did not support prompts. With this change, prompts can now be discovered and retrieved through the Lambda handler, bringing feature parity with MCP servers outside Lambda.

### Changes

* Added prompt registration support in `MCPLambdaHandler`.
* Implemented handlers for:

  * `initialize` → now includes prompt discovery.
  * `prompts/list` → returns available prompts.
  * `prompts/get` → retrieves prompt details.
* Ensured compatibility with existing MCP Lambda functionality.

### User experience

**Before:**

* MCP servers on Lambda did not support prompts.
* No way to list or fetch prompts from Lambda-hosted MCP servers.

**After:**

* MCP servers on Lambda now fully support prompt discovery and retrieval (`initialize`, `prompts/list`, `prompts/get`).
* Developers can use prompts seamlessly when deploying MCP servers with `MCPLambdaHandler`.

## Checklist

* [x] I have reviewed the [[contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

Is this a breaking change? (Y/N)
**N**

**RFC issue number**:  #731 

## Integration Testing:

Deploying the MCP server on AWS Lambda:

<img width="1738" height="846" alt="Screenshot 2025-11-01 020752" src="https://github.com/user-attachments/assets/10ed9f52-859f-42da-8a93-fb0458c05be8" />

Testing the Prompts on MCP Inspector:

<img width="1841" height="930" alt="Screenshot 2025-11-01 020915" src="https://github.com/user-attachments/assets/e10634ee-3561-41ac-a312-42bdddddcb09" />

-----------

<img width="1854" height="943" alt="Screenshot 2025-11-01 020843" src="https://github.com/user-attachments/assets/5d29404b-158f-4c3c-aa74-8a137f4e4cb9" />

Testing the prompts features on Cursor:

<img width="678" height="913" alt="image" src="https://github.com/user-attachments/assets/90b378b7-f39f-49fc-82b9-d8e95625df5a" />

-----------

<img width="678" height="253" alt="image" src="https://github.com/user-attachments/assets/3736571a-aa43-4a48-9463-95027095b3d6" />



Checklist:

* [x] Migration process documented
* [x] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

Dan-Arnin committed:
feat: add prompt registration and rendering functionality to MCPLambdaHandler

Dan-Arnin committed:
docs: add prompts section to README with examples and usage instructions

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'awslabs:main' into lambda_mcp_prompts

Dan-Arnin commented:
@Lukas-Xue, @mikegc-aws awaiting your review on this PR. Thanks!

krokoko committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
feat: add prompt decorator tests for registration, naming, and handling

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
@Lukas-Xue @mikegc-aws  - unit test cases have been added as requested. Integration testing proofs have also been added in the PR Summary for your perusal. Please review it and let me know your thoughts.

Thanks in advance!

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Lukas-Xue commented:
Thanks Dan, this looks great with comprehensive testing. Let's merge it!

Lukas-Xue reviewed (dismissed):
LGTM

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
Huge thanks @Lukas-Xue for reviewing the code and approving it. 

@mikegc-aws awaiting your response. Thanks in advance.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
@mikegc-aws , @scottschreckengaust - could you please review the PR if you don't mind. It has been open for quite a long time now.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

scottschreckengaust committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
@scottschreckengaust @mikegc-aws it would be wonderful if you could review the code and approve the same. Thanks in advance.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

mikegc-aws reviewed (dismissed):
LGTM. I ran the full test suite locally (100 tests passed) and reviewed the implementation against the MCP prompts spec — behavior and shapes look correct, and this is backward compatible.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
Any updates? It would be helpful if you could get this branch merged now that all approvals are in place.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
chore: ran precommit for safe merge

Dan-Arnin committed:
chore: added a newline at eod of readme for precommit

Dan-Arnin commented:
@Lukas-Xue @mikegc-aws precommit was failing, pushed a few commits to fix the same, would be helpful if you could review the new changes and approve the same, thanks in advance

mikegc-aws commented:
Hey @Dan-Arnin — thanks for your patience on this and for keeping the branch up to date. The overall approach is solid and the test coverage is great. I have a few items I'd like addressed before we can merge:

**Must-fix:**

1. **Missing `required` field on prompt arguments** — The MCP `PromptArgument` schema includes a `required: bool` field. You can use `inspect.signature(f)` to check whether each parameter has a default value — if it does, `required` should be `false`, otherwise `true`. This is important for clients to know which arguments are mandatory.

2. **Session validation** — `prompts/list` and `prompts/get` don't validate that a session exists, but all other method handlers (`tools/list`, `tools/call`, `resources/list`, `resources/read`) do. Please add the same session check for consistency and correctness per the MCP lifecycle.

3. **Argument type coercion** — `_render_prompt` passes argument values through as-is from the JSON-RPC params. If a prompt function expects `count: int` but receives the string `"42"`, it won't be coerced. The tool handler already does type coercion — please apply the same pattern here.

**Nice-to-have (non-blocking):**

- Consider supporting multi-message prompt responses (list of messages with roles), since the MCP spec allows it. A simple approach: if the function returns a `list`, treat each item as a message; if it returns a `str`, wrap it as a single user message as you do today.
- The auto camelCase naming (`generate_code` → `generateCode`) differs from FastMCP's convention (which keeps snake_case). Worth noting in the README so users aren't surprised.

These fixes should be straightforward. Happy to re-approve once they're in. Thanks again for the contribution!

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
fix(mcp-lambda-handler): prompt required field, session validation, type coercion, multi-message support

- Add required field to prompt args via inspect.signature
- Add session validation to prompts/list and prompts/get
- Coerce string args to correct types in _render_prompt
- Support list return from prompts as multi-message response
- Add tests and README note on camelCase prompt naming

Dan-Arnin committed:
Merge branch 'lambda_mcp_prompts' of https://github.com/Dan-Arnin/mcp into lambda_mcp_prompts

Dan-Arnin committed:
chore: ran pre-commit to make sure files are compliant

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin commented:
@mikegc-aws  @Lukas-Xue  it would be wonderful if you could review the changes.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

mikegchambers reviewed (commented):
LGTM - all three requested changes from my March review are addressed (required field inference, session validation, type coercion). Tests are comprehensive and the decorator pattern is consistent with the existing tool/resource handlers.

One heads-up: PR #3284 (protocol upgrade to 2025-11-25) changes the capabilities format from `{'list': True, 'get': True}` to `{'listChanged': False}`. When both PRs land, whoever merges second will need a quick capabilities-shape update for prompts. Trivial conflict but worth noting for merge order.

Thanks for your patience on this one, Dan - great contribution.

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts

Dan-Arnin committed:
Merge branch 'main' into lambda_mcp_prompts
